packages/reputation-miner/test/MaliciousReputationMinerWrongJRHRightNNodes.js

```diff
@@ -0,0 +1,82 @@
+import ReputationMinerTestWrapper from "./ReputationMinerTestWrapper";
+import PatriciaTreeNoHash from "../patriciaNoHashKey";
+
+const ethers = require("ethers");
+
+const miningCycleDuration = ethers.utils.bigNumberify(60).mul(60).mul(24); // 24 hours
+const minStake = ethers.utils.bigNumberify(10).pow(18).mul(2000);
+const constant = ethers.utils.bigNumberify(2).pow(256).sub(1).div(miningCycleDuration);
+
+class MaliciousReputationMinerWrongJRHRightNNodes extends ReputationMinerTestWrapper {
+  // Only difference between this and the 'real' client should be that it submits a bad JRH
+
+  constructor(opts, entriesToFalsify, entriesToSkip) {
+    super(opts);
+    this.entriesToFalsify = entriesToFalsify.map(x => x.toString());
+    this.entriesToSkip = entriesToSkip.map(x=>x.toString());
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  async respondToChallenge() {
+    // This client sometimes won't be able to respond to challenge - we mess up its JRH with a hash it doesn't know about
+  }
+
+  async submitRootHash(startIndex = 1) {
+    const hash = await this.getRootHash();
+    const repCycle = await this.getActiveRepCycle();
+    // Get how much we've staked, and thefore how many entries we have
+    let entryIndex;
+    const [, balance] = await this.tokenLocking.getUserLock(this.clnyAddress, this.minerAddress);
+    const reputationMiningWindowOpenTimestamp = await repCycle.getReputationMiningWindowOpenTimestamp();
+    for (let i = ethers.utils.bigNumberify(startIndex); i.lte(balance.div(minStake)); i = i.add(1)) {
+      // Iterate over entries until we find one that passes
+      const entryHash = await repCycle.getEntryHash(this.minerAddress, i, hash);
+      const block = await this.realProvider.getBlock("latest");
+      const { timestamp } = block;
+
+      const target = ethers.utils
+        .bigNumberify(timestamp)
+        .sub(reputationMiningWindowOpenTimestamp)
+        .mul(constant);
+
+      if (ethers.utils.bigNumberify(entryHash).lt(target)) {
+        entryIndex = i;
+        break;
+      }
+    }
+    if (!entryIndex) {
+      return new Error("No valid entry for submission found");
+    }
```

Comment:
I feel like all of the above could probably be extracted into a `miner-utils.js` file somewhere.

Reply:
The relevant functionality is already in `getEntryIndex` in the class. The miner I copied this from just hadn't been updated since that refactoring. Both this file and that file are a lot smaller now they're using that function 😄